src/Core/Config/Middleware/ExtensionMiddleware.php

```diff
@@ -66,10 +67,14 @@ protected function getExtraConfig($class, $classConfig, $excludeMiddleware)
             list($extensionClass, $extensionArgs) = ClassInfo::parse_class_spec($extension);
             // Strip service name specifier
             $extensionClass = strtok($extensionClass, '.');
-            if (!class_exists($extensionClass)) {
+            if (!class_exists($extensionClass) && !Injector::inst()->has($extensionClass)) {
                 throw new InvalidArgumentException("$class references nonexistent $extensionClass in 'extensions'");
             }
 
+            if (!class_exists($extensionClass)) {
```

Comment:
Maybe you could combine this into the previous if statement to reduce duplicating the condition

Reply:
I thought about it, but it gets difficult as the previous condition is dealing specifically with the lack of reference. It could be done, however the trade-off would be a touch of pyramid nesting I think.

Reply:
How about something like this?

```php
$extensionInstance = $this->getExtensionInstance($extensionClass);

protected function getExtensionInstance($className)
{
    if (Injector::inst()->has($className)) {
        return Injector::inst()->get($className);
    }

    if (!class_exists($className)) {
        throw new Exception('...');
    }

    return $className;
}
```

Something like this might be good, you could implement it in the places where the logic is duplicated.

Also note that returning either an instance or a class name string from the logic seems odd (before this PR it only returns a class name string), so perhaps you could change it to always return an instance of the extension

Reply:
Yeah sweet I'll do something like that :)

Reply:
So I looked at this - but in the end there are two issues:
1. The exceptions require `$class` (that the extension would belong to) which would not exist if I abstracted the method - passing it around seems silly. I had the new method setting on Extension because it seemed the only logical place for it to be, otherwise I'd introduce coupling between the `trait` and the `Middleware`. One of the calls actually results in an `error` as opposed to an `Exception` as well... so there'd need to be code to allow for that anyway, otherwise we'd be changing the API.
2. Adding functions frivolously for what essentially amounts to 'Injector::doOneThing()` also seemed silly at that point. I could add `getConcreteClassName` to `Injector`... `getServiceNamedSpec` already exists, but it's protected.

Reply:
I was going to recommend exactly what @robbieaverill suggested. That is, using has() instead of the spec. You can do it inline though. It doesn't need to be a method.

Reply:
That does not consider the code outside this change, where `$extensionClass` is then used to look up the concrete parent chain via `ClassInfo` - which would severely break if it was not both set, or a valid concrete class.

Reply:
If you just rely on injector you don't need to do class_exists() or other checks. Injector will internally validate the service is valid. Do you really need to do concrete class lookups outside of injector?

Reply:
Yes.
https://github.com/silverstripe/silverstripe-framework/blob/d75ce9b39ef4803a651b66f873fdbe5f48cd2ff2/src/Core/Config/Middleware/ExtensionMiddleware.php#L99-L109